avocado/core/nrunner.py

```diff
@@ -613,97 +570,194 @@ def wait(self):
+            description = self.PROG_DESCRIPTION
+        self._setup_parser(prog, description)
+
+    def _setup_parser(self, prog, description):
+        self.parser = argparse.ArgumentParser(prog=prog,
+                                              description=description)
+        subcommands = self.parser.add_subparsers(dest='subcommand')
+        subcommands.required = True
+        for cmd_meth in self._get_commands_method_without_prefix():
+            attr = "CMD_%s_ARGS" % cmd_meth.upper()
+            cmd = cmd_meth.replace('_', '-')
+            cmd_parser = subcommands.add_parser(cmd)
+            if hasattr(self, attr):
+                for arg in getattr(self, attr):
+                    cmd_parser.add_argument(*arg[0], **arg[1])
+
+    def _get_commands_method_without_prefix(self):
+        prefix = 'command_'
+        return [c[0][len(prefix):]
+                for c in inspect.getmembers(self, inspect.ismethod)
+                if c[0].startswith(prefix)]
+
+    def run(self):
+        """
+        Runs the application by finding a suitable command method to call
+        """
+        args = vars(self.parser.parse_args())
+        subcommand = args.get('subcommand')
+        if subcommand in self.get_commands():
+            meth_name = 'command_' + subcommand.replace('-', '_')
+            if hasattr(self, meth_name):
+                kallable = getattr(self, meth_name)
+                return kallable(args)
+
+    def get_commands(self):
+        """
+        Return the command names, as seen on the command line application
+
+        For every method whose name starts with "command_", a the name of
+        the command follows, with underscores replaced by dashes.  So, a
+        method named "command_foo_bar", will be a command available on the
+        command line as "foo-bar".
+
+        :rtype: list
+        """
+        return [c.replace('_', '-') for c in
+                self._get_commands_method_without_prefix()]
+
+    def get_capabilities(self):
+        """
+        Returns the runner capabilities, including runnables and commands
+
+        This can be used by higher level tools, such as the entity spawning
+        runners, to know which runner can be used to handle each runnable
+        type.
+
+        :rtype: dict
+        """
+        return {"runnables": [k for k in self.RUNNABLE_KINDS_CAPABLE.keys()],
+                "commands": self.get_commands()}
+
+    def get_runner_from_runnable(self, runnable):
+        """
+        Returns a runner that is suitable to run the given runnable
+
+        :rtype: instance of class inheriting from :class:`BaseRunner`
+        :raises: ValueError if runnable is now supported
+        """
+        runner = self.RUNNABLE_KINDS_CAPABLE.get(runnable.kind, None)
+        if runner is not None:
+            return runner(runnable)
+        raise ValueError('Unsupported kind of runnable: %s' % runnable.kind)
+
+    def command_capabilities(self, _):
```

Comment:
@clebergnu 

I tried to explain my point here https://github.com/avocado-framework/avocado/pull/3614#discussion_r386549320

Reply:
I answered it here:

https://github.com/avocado-framework/avocado/pull/3614#issuecomment-593549641

But because I replied to an email, it showed up in the wrong place.  Let me know if that explains my point.